### PR TITLE
ZK-3838: users can paste non-digital characters into a intbox

### PR DESCRIPTION
**Pull request by klyvechen**

### CODE REVIEW

File: zul/src/archive/web/js/zul/inp/InputWidget.js [hunk 1]
```diff
@@ -678,13 +678,32 @@ zul.inp.InputWidget = zk.$extends(zul.Widget, {
...
+				window.clipboardData.getData('text'),
+				minusIndex = keys.indexOf('-');
+			if (minusIndex != -1) {
+				keys = keys.substring(0,minusIndex) + '\\' + keys.substring(minusIndex);
+			}
+			persentIndex = keys.indexOf('%');
+			if (persentIndex != -1) {
+				keys = keys.substring(0,persentIndex) + '/' + keys.substring(persentIndex);
+			}
+			var regKey = '^[' + keys + ']+$';
```
Comment (scribetw) [inline zul/src/archive/web/js/zul/inp/InputWidget.js]: Don't escape yourself

Use https://stackoverflow.com/a/9310752

File: zul/src/archive/web/js/zul/inp/InputWidget.js [hunk 2]
```diff
@@ -678,13 +678,32 @@ zul.inp.InputWidget = zk.$extends(zul.Widget, {
...
+			}
+		} else {
+			var orgEvtText = evt.domEvent.originalEvent.clipboardData.getData('text'),
+				text = (orgEvtText) ? orgEvtText :
+				window.clipboardData.getData('text'),
+				minusIndex = keys.indexOf('-');
+			if (minusIndex != -1) {
+				keys = keys.substring(0,minusIndex) + '\\' + keys.substring(minusIndex);
+			}
+			persentIndex = keys.indexOf('%');
```
[the remark this answers is not in the document]
Comment (klyvechen) [reply zul/src/archive/web/js/zul/inp/InputWidget.js]: ok

File: zk/src/archive/web/js/zk/widget.js [hunk 3]
```diff
@@ -4329,6 +4329,26 @@ wgt.setListeners({
 			if (p) p.doKeyPress_(evt);
 		}
 	},
+	/** Called when the user paste text when this widget has the focus ({@link #focus}).
+	 * <p>Notice that not every widget can have the focus.
+	 * A widget doesn't need to listen the past DOM event.
```
[the remark this answers is not in the document]
Comment (klyvechen) [reply zk/src/archive/web/js/zk/widget.js]: ok

File: zktest/src/archive/test2/B85-ZK-3838.zul [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+        <textbox id ="textbox2" value="123"/>
+    </window>
+    avoid pasting non-number characters
+    <intbox/>
+
+
+
+    <label  multiline="true">
+        1. The intbox can limit the input even though by paste method.
+        2. Copy the value in label1 and paset it to intbox.
```
Comment (scribetw) [inline zktest/src/archive/test2/B85-ZK-3838.zul]: replace all "paset" -> paste

File: zktest/src/archive/test2/B85-ZK-3838.zul [hunk 5]
```diff
@@ -0,0 +1,25 @@
+<zk xmlns:h="client/attribute">
```
Comment (scribetw) [inline zktest/src/archive/test2/B85-ZK-3838.zul]: Non used xmlns, delete it

File: zktest/src/archive/test2/config.properties [hunk 6]
```diff
@@ -2450,6 +2450,7 @@ B85-ZK-3848.zul=A,E,DefinitionLoader,land-addon
 ##zats##B85-ZK-3691.zul=A,E,Listbox,groupSelect,Listgroup,Listgroupfoot,_selectable
 ##zats##B85-ZK-3910.zul=A,E,ckeditor,iframe,remove
 ##zats##B85-ZK-3698.zul=A,E,Decimalbox,value,NPE
+B85-ZK-3698.zul=A,E,InputWidegt,copy,paste,string,number,IE9
```
Comment (scribetw) [inline zktest/src/archive/test2/config.properties]: 1. B85-ZK-3838
2. InputWidget

File: zul/src/archive/web/js/zul/inp/InputWidget.js [hunk 7]
```diff
@@ -678,13 +678,29 @@ zul.inp.InputWidget = zk.$extends(zul.Widget, {
...
+					return true;
+				}
+			}
+		} else {
+			var orgEvtText = evt.domEvent.originalEvent.clipboardData.getData('text'),
+				text = (orgEvtText) ? orgEvtText :
+				window.clipboardData.getData('text'),
+				minusIndex = keys.indexOf('-');
+			keys = escapeRegExp(keys);
+			var regKey = '^[' + keys + ']+$';
```
Comment (scribetw) [inline zul/src/archive/web/js/zul/inp/InputWidget.js]: The method is not widely used, so just inline it
```js
var regKey = '^[' + keys.replace(/[-[\]{}()*+?.,\\^$|#\s]/g, '\\$&') + ']+$';
```

File: zul/src/archive/web/js/zul/inp/InputWidget.js [hunk 8]
```diff
@@ -678,13 +678,29 @@ zul.inp.InputWidget = zk.$extends(zul.Widget, {
...
+				&& keys.indexOf(String.fromCharCode(code)) < 0) {
+					evt.stop();
+					return true;
+				}
+			}
+		} else {
+			var orgEvtText = evt.domEvent.originalEvent.clipboardData.getData('text'),
+				text = (orgEvtText) ? orgEvtText :
+				window.clipboardData.getData('text'),
+				minusIndex = keys.indexOf('-');
```
Comment (scribetw) [inline zul/src/archive/web/js/zul/inp/InputWidget.js]: Non used var

File: zul/src/archive/web/js/zul/inp/InputWidget.js [hunk 9]
```diff
@@ -678,13 +678,29 @@ zul.inp.InputWidget = zk.$extends(zul.Widget, {
...
+				if (!evt.altKey && !evt.ctrlKey && _keyIgnorable(code)
+				&& keys.indexOf(String.fromCharCode(code)) < 0) {
+					evt.stop();
+					return true;
+				}
+			}
+		} else {
+			var orgEvtText = evt.domEvent.originalEvent.clipboardData.getData('text'),
+				text = (orgEvtText) ? orgEvtText :
+				window.clipboardData.getData('text'),
```
Comment (scribetw) [inline zul/src/archive/web/js/zul/inp/InputWidget.js]: Just one line, don't break
```js
text = (orgEvtText) ? orgEvtText : window.clipboardData.getData('text'),
```

File: zktest/src/archive/test2/B85-ZK-3838.zul (hunk 4 above)
Comment (klyvechen) [reply zktest/src/archive/test2/B85-ZK-3838.zul]: ok

File: zktest/src/archive/test2/B85-ZK-3838.zul (hunk 5 above)
Comment (klyvechen) [reply zktest/src/archive/test2/B85-ZK-3838.zul]: ok

File: zul/src/archive/web/js/zul/inp/InputWidget.js (hunk 7 above)
Comment (klyvechen) [reply zul/src/archive/web/js/zul/inp/InputWidget.js]: ok

File: zktest/src/archive/test2/config.properties (hunk 6 above)
Comment (klyvechen) [reply zktest/src/archive/test2/config.properties]: ok

File: zul/src/archive/web/js/zul/inp/InputWidget.js (hunk 9 above)
Comment (klyvechen) [reply zul/src/archive/web/js/zul/inp/InputWidget.js]: ok

File: zul/src/archive/web/js/zul/inp/InputWidget.js (hunk 8 above)
Comment (klyvechen) [reply zul/src/archive/web/js/zul/inp/InputWidget.js]: ok

**Comment by scribetw** [comment]
@zkadmin rebase

**Comment by zkadmin** [comment]
I just pushed up the changes, enjoy!

**Pull request merged**